README
$Id: README,v 1.5 2004/12/16 10:13:26 mamezo Exp $
--
    
Xprobe2 is a remote active operating system fingerprinting tool.
Xprobe2 uses advanced technics, such as the usage of fuzzy logic
to match fingerprinting results with the tool's signature database,
unique fingerprinting methods and other, to provide with accurate
results.

Xprobe2 has evolved from the original Xprobe tool, but it has been
completely rewritten from scratch in C++.



License
-------
Xprobe2 is distributed under GNU license. See COPYING for more details.

The fingerprinting database is is available for free use by open source 
software under the terms of the GNU General Public License.

For commercial usage please contact: ofir@sys-security.com
    
        
Requirements
------------

- A C++ compiler. (gcc will do)

- libusi++ (included with the distribution)

- lipcap (use version 0.6.x or later)
  If libpcap is not installed in standard path, use --with-libpcap-libraries=/path/ 
  and --with-libcap-includes=/path options.


Supported platforms
-------------------

The tool has been succesefully compiled and tested under the following platforms:

FreeBSD 4.x, 5.x 
Linux 2.0.x, 2.2.x, 2.4.x
Solaris 2.x
OpenBSD 2.x, 3.x
NetBSD 1.4.x, 1.5.x, 1.6.x
IRIX (with SGI freeware libpcap, http://freeware.sgi.com/, naitive SGI compiler).


Platforms which we are able to fingerprint
------------------------------------------

see etc/xprobe2.conf


How to install
---------------

tar xvfz xprobe2-{release}.tar.gz
cd xprobe-{release}
./configure
(or ./configure --with-libpcap-libraries=/usr/local/lib --with-libcap-includes=/usr/local/include)
make
make install

send complains to fygrave@tigerteam.net if the compilation breaks. (use
--enable-debug to track/report errors)


How to use:
-----------

See manual for details or use the -h command line to list available options.


Architecture
------------

Xprobe2 consists of 2 major parts: core fingerprinting engine which
includes fuzzy signatures matching engine, signatures processing,
and is generally intrefacing between the modules, and tests, which
are presented as (extrenal) dynamically loadable shared modules.

Core module has no idea about the signatures nor how they are being
applied to received packets. Please see fuzzy_fingerprinting paper in
docs for details.    

Xprobe2 modules are supposed to provide routines for signature element
parsing for each module, module names, initialization routines, module
execution routines, module deinitialization signatures and other
specific information. Please see docs/modules_howto.txt for details.


Docs & Updates
--------------

http://www.sys-security.com/html/projects/X.html


Other related webpages
----------------------

http://o0o.nu/sec/xprobe/
http://www.sourceforge.net/projects/xprobe/
http://xprobe.sourceforge.net/


Where to mail bugs/questions/ideas/patches/fixes:
-------------------------------------------------

 Fyodor Yarochkin <fygrave@o0o.nu>         Ofir Arkin <ofir@sys-security.com>
 http://o0o.nu                             The Sys-Security Group
                                           http://www.sys-security.com
 
 Meder Kydyraliev <meder@o0o.nu>
                                 
Flames
-------

/dev/null                                   


CHANGELOG
2005-07-26  mederchik

	* man page updated with example on using SNMP module

2005-07-26  mederchik

	* several sample snmp_sysdescr substrings for freebsd and mac os

2005-07-26  mederchik

	* finished SNMPv2c module, registers keyword "snmp_sysdescr", which
	should be a substring of a sysDescr.0 string

2005-07-26  mederchik

	* fix for port_is_open() method to correctly test whether given
	port is open

2005-07-21  mederchik

	* a list of community strings to try is in xprobe2.conf file

2005-07-21  mederchik

	* skeleton for the SNMP module

2005-07-21  mederchik

	* a slight change in logic of the module, now it send NBT node
	status query before session setup packet

2005-07-18  mederchik

	* base for SNMP module

2005-07-18  mederchik

	* fixed several bugs and added more sanity checks. Also added NBT
	node status query to get the name remote box listens on (also gets
	us MAC).

2005-07-18  mederchik

	* added OS_Matrix::add_result(XPROBE_MATCH_NO) calls when no
	matches are found

2005-07-18  mederchik

	* added OpenUDPSocket routine

2005-07-18  mederchik

	* fixed bug when using Xprobe_Module_Params INTLIST. The bug showed
	up after we added possibility to have variable number of keywords
	for fingerprints and resulted in extra OS_Matrix::add_result calls

2005-07-18  mederchik

	* base for SNMP module

2005-07-18  mederchik

	* update from Ofir (keywords for smb module)

2005-06-26  mederchik

	* added support for OS_Matrix to be able to still give 100%
	whenever different fingerprints have different number of keywords

2005-06-26  mederchik

	* several updates to the modules: -added new checkpoint value for
	TTL singature generation (60); -added sanity check to the SMB
	module;

2005-06-26  mederchik

	* minor updates to the man page and howtos

2005-06-23  mederchik

	* new SMB module import. Module extract Native OS and Native Lanman
	from session setup and X reply. Module supports both SMB over
	NBT(139) and naked SMB(445). Module will be executed if either 139
	or 445 TCP port is open

2005-06-23  mederchik

	* util func to compare strings

2005-06-23  mederchik

	* added method to check if specific port is open

2005-06-20  mederchik

	* updated TODO

2005-06-19  mederchik

	* update from Ofir

2005-06-17  mederchik

	* removed debugging printf()

2005-06-17  mederchik

	* CREDITS

2005-06-17  mederchik

	* small fix for compiler warning on osx

2005-06-17  mederchik

	* fixes for bugs reported by Alf B Lervaag <alfborge@stud.ntnu.no>

2005-06-09  mederchik

	* update from Ofir

2005-02-16  mederchik

	* stupid bug in if statment, resulted in 97% match, as border value
	of echoed_dtsize was equal 64

2005-02-15  mederchik

	* updated CHANGELOG

2005-02-15  mederchik

	* new default port 65534

2005-02-15  mederchik

	* fixed to_string as it displayed the same ip because of
	inet_ntoa() call

2005-02-15  mederchik

	* upate from Ofir

2005-02-14  mederchik

	* update man page with info on -A and add some checks for proper
	use of -A

2005-02-14  mederchik

	* experimental suport for detection of transparent proxies,
	firewalls/NIDSs spoofing RSTs on behalf of target added to
	portscanner, to enable use -T in conjunction with -A

2005-02-14  mederchik

	* fixed singature generation and matching: echoed_dtsize of
	icmp_unreach can now have arbitrary number, ip_id_strategy
	generation in tcp_rst now matches <=256 instead of <256 for I

2005-02-14  mederchik

	* added int_to_string() to xp_lib

2005-02-14  mederchik

	* nicer ouput of packets

2005-02-14  mederchik

	* signature update from Ofir including updated fingerprints for new
	module

2005-02-13  mederchik

	* employ new usi++ methods

2005-02-13  mederchik

	* usi++ modifications for packet comparison that will be used by
	portscanner to detect variations in replies from target

2005-02-13  mederchik

	* updated port list for -B in example

2005-02-13  mederchik

	* cosmetic change to be consistent with xprobe2 output

2005-02-13  mederchik

	* fixed bug when configure for USI++ was run two times; added some
	symbols to configure.in to port to MacOSX

2005-02-09  mederchik

	* automatically generated CHANGELOG

2005-02-09  mederchik

	* fixed ICMP info module disabling issue

2005-02-09  mederchik

	* added new keywords, added some minor optimizations, fixes and
	minor enhacements

2005-02-09  mederchik

	* added operator=, as default was copying pointers

2005-02-09  mederchik

	* updated from Ofir + added full FreeBSD 5.3 fingerprint with and
	without rfc 13223 extensions enabled

2005-02-09  mederchik

	* start using CREDITS for credits

2005-02-09  mederchik

	* commited fixes from Kurt Congdon <kurt@vdubgeek.net>

2005-02-09  mederchik

	* New TCP RST module added

2005-01-14  mederchik

	* fix bug reported by Kurt Congdon <kurt@vdubgeek.net>:
	Target_Net::getnext() returned special addresses (network num and
	broadcast)

2005-01-12  mederchik

	* signature update from Ofir, forget to remove a var from
	xp_lib.cc, update copyright to 2005

2005-01-12  mederchik

	* fixed non-zero ACK in SYN packet

2005-01-12  mederchik

	* remove debugging statements

2005-01-12  mederchik

	* new xprobe_module_param type introduced _INTLIST, which handles
	comma separated list of integers as value

2005-01-12  mederchik

	* regenerated configure script, got rid of Wmissing-prototypes, as
	it was causing warnings

2005-01-12  mederchik

	* move tokenize to a xp_lib class that from now on will be holding
	all utility methods/funcs. TODO: move all xp_ funcs there too

2005-01-10  mederchik

	* new configure script for usi

2005-01-10  mederchik

	* port 6000 added to blind guess list

2005-01-10  mederchik

	* got duplicate sigs from somwhere, fixed

2005-01-05  mamezo

	* sigs update

2004-12-27  mederchik

	* Richard Atterer <atterer@debian.org> pointed to old config.guess
	and config.sub, update to new version

2004-12-27  mederchik

	* get rid of carriage return char

2004-12-20  mederchik

	* update from Ofir


src/defines.h
/* $Id: defines.h.in,v 1.17 2005/07/18 11:08:24 mederchik Exp $ */
/*
** Copyright (C) 2001 Fyodor Yarochkin <fygrave@tigerteam.net>,
**                    Ofir Arkin       <ofir@sys-security.com>
**
** This program is free software; you can redistribute it and/or modify
** it under the terms of the GNU General Public License as published by
** the Free Software Foundation; either version 2 of the License, or
** (at your option) any later version.
**
**
** This program is distributed in the hope that it will be useful,
** but WITHOUT ANY WARRANTY; without even the implied warranty of
** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
** GNU General Public License for more details.
**
** You should have received a copy of the GNU General Public License
** along with this program; if not, write to the Free Software
** Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
*/

#ifndef DEFINES_H
#define DEFINES_H

#define VERSION "0.3"
#define BANNER \
"\nXprobe2 v."VERSION\
" Copyright (c) 2002-2005 fyodor@o0o.nu, ofir@sys-security.com, meder@o0o.nu"\
"\n"

#define DATA_SIZE 20
#define DEF_TIMEOUT 10
#define DEF_SEND_DELAY 0.01
#define PACKBUF_SIZ 1024
#define DEFAULT_CONFIG "/etc/xprobe2/xprobe2.conf"
#define DEFAULT_MATCHES 10

#define UDP_PORT 32132
#define UDP_DATA_SIZE 70
#define UDP_LEN_ALL (sizeof(struct ip) + sizeof(struct udphdr) + UDP_DATA_SIZE)

#define PING_PAYLOAD_SIZE 56

#define TEST_RESULT_BROKEN -1
#define TEST_RESULT_ZERO    0
#define TEST_RESULT_GOOD    1
#define TEST_RESULT_BAD     2

#define UDP_CKSUM_BADDGRAM TEST_RESULT_BROKEN
#define UDP_CKSUM_ZERO      TEST_RESULT_ZERO
#define UDP_CKSUM_BAD       TEST_RESULT_BAD
#define UDP_CKSUM_GOOD      TEST_RESULT_GOOD

#define IP_CKSUM_ZERO      TEST_RESULT_ZERO
#define IP_CKSUM_BAD       TEST_RESULT_BAD
#define IP_CKSUM_GOOD      TEST_RESULT_GOOD

#define IP_ID_ZERO          TEST_RESULT_ZERO
#define IP_ID_BAD           TEST_RESULT_BAD
#define IP_ID_GOOD          TEST_RESULT_GOOD
#define IP_ID_FLIPPED       (TEST_RESULT_BAD + 1)

/* we don't expect packets bigger than 1500 */
#define LPCAP_SNAPLEN       1500
#define LPCAP_TIMEOUT       2000

/* return values for ICMP_UNREACH length check */

#define ICMPUNREACH_LEN_OK  1
#define ICMPUNREACH_LEN_GT  2
#define ICMPUNREACH_LEN_LS  3

/* return values for frag bits test */

#define FRAG_BITS_OK        0
#define FRAG_BITS_FLIPPED   1
#define FRAG_BITS_ZERO      2

/* TTL_DELTA for fuzzy TTL matching */

#define TTL_DELTA			25

/* defines for tcp_ports, udp_ports and protocols maps */

#define XPROBE_TARGETP_OPEN 1
#define XPROBE_TARGETP_CLOSED 2
#define XPROBE_TARGETP_FILTERED 3


#define MAXIFNUM 64

/* DSO modules stuff */

#if defined(LINUX) || defined(__FreeBSD__)
#define DLOPENFLAGS     RTLD_LAZY | RTLD_GLOBAL
#define DLSYMPREFIX    ""
#else
#define DLOPENFLAGS     RTLD_LAZY
#define DLSYMPREFIX    "_"
#endif

#define N_TESTNAME "test_name"
#define N_TESTTYPE "test_type"
#define N_TEST_INIT "test_init"
#define N_TEST_RUN "test_run"
#define N_TEST_FINI "test_fini"

#define XPROBE_MODULE_ENABLED 0
#define XPROBE_MODULE_DISABLED 1

/* some misc stuff */

#define FAIL		-1
#define OK			0
#define RETRY		-2
#define TRUE		1 
#define FALSE		0

#define XPROBE_DEBUG_ALL          0xffffffff
#define XPROBE_DEBUG_SIGNATURES   1
#define XPROBE_DEBUG_DSOTESTS     2
#define XPROBE_DEBUG_INIT         4
#define XPROBE_DEBUG_MODULES      8
#define XPROBE_DEBUG_OSMATRIX    16  
#define XPROBE_DEBUG_CONFIG      32
#define XPROBE_DEBUG_TARGET      64

#define XPROBELOG_XP_SESS_START     1
#define XPROBELOG_MSG_RUN           2
#define XPROBELOG_MOD_SESS_START    3
#define XPROBELOG_MSG_MODULE        4
#define XPROBELOG_MOD_SESS_END      5
#define XPROBELOG_TG_SESS_START     6
#define XPROBELOG_REACH_SESS_START  7
#define XPROBELOG_MSG_STATE         8
#define XPROBELOG_MSG_RTT           9
#define XPROBELOG_REACH_SESS_END    10
#define XPROBELOG_INFO_SESS_START   11
#define XPROBELOG_PS_SESS_START     12
#define XPROBELOG_STATS_SESS_START  13
#define XPROBELOG_MSG_PS_TCPST      14
#define XPROBELOG_MSG_PS_UDPST      15
#define XPROBELOG_STATS_SESS_END    16
#define XPROBELOG_PSDET_SESS_START  17
#define XPROBELOG_MSG_PORT          18
#define XPROBELOG_PSDET_SESS_END    19
#define XPROBELOG_PS_SESS_END       20
#define XPROBELOG_INFO_SESS_END     21
#define XPROBELOG_GUESS_SESS_START  22
#define XPROBELOG_MSG_PRIMARY       23
#define XPROBELOG_MSG_SECONDARY     24
#define XPROBELOG_GUESS_SESS_END    25
#define XPROBELOG_TG_SESS_END       26
#define XPROBELOG_XP_SESS_END       27
#define XPROBELOG_OTHER_TCPP		28
#define XPROBELOG_OTHER_UDPP		29

#ifndef DEFAULT_DEBUG_LEVEL
#define DEFAULT_DEBUG_LEVEL 0
#endif

#define xprobe_debug(level, fmt, args...) ui->debug(level,\
__FILE__, __LINE__, fmt, args)

#define xprobe_mdebug(level, str) ui->debug(level,\
__FILE__, __LINE__, str)

#endif /* DEFINES_H */
